File: packages/quran-engine-go/batch6.go
package quranengine

// The corpora added upstream in Al-Islam 4.6.5: the 99 Names in depth, and the chains of
// transmission of the Ten Readings. See ../../docs/23-names-depth.md and 24-isnad.md.

import (
	"sort"
	"strings"
)

// ---- the Names in depth ----------------------------------------------------------

// NameTheme is one of the nine themes the Names are grouped under.
type NameTheme struct {
	ID    string `json:"id"`
	Label string `json:"label"`
}

// NameOccurrence is one ayah a Name appears in.
//
// Token is a pointer because the corpus could not place ten of them in their ayah's tokens: the
// ayah is still right, so a consumer shows the verse whole and highlights nothing.
type NameOccurrence struct {
	Surah  int  `json:"surah"`
	Ayah   int  `json:"ayah"`
	Token  *int `json:"token"`
	Tokens int  `json:"tokens"`
}

// NameDepth is the layer under names-of-allah.json: what the Name is built on and what it asks
// of a reader.
//
// The written material is Tilawa's (Jamil Hammoudeh), used with permission; the occurrences
// point into this engine's own Hafs text.
type NameDepth struct {
	Number int `json:"number"`
	// Root prints SPACED, the way a grammar book sets it ("ر ح م"). RootKey closes it up to the
	// form morphology.json stores.
	Root        string           `json:"root"`
	Theme       string           `json:"theme"`
	Explanation string           `json:"explanation"`
	Living      string           `json:"living"`
	Occurrences []NameOccurrence `json:"occurrences"`
}

type namesDepthFile struct {
	Themes []NameTheme `json:"themes"`
	Names  []NameDepth `json:"names"`
}

// RootKey returns a spaced root as morphology stores it: "ر ح م" -> "رحم".
//
// Every port strips whitespace explicitly rather than trimming, because a root is spaced in the
// middle and not only at the ends.
func RootKey(root string) string {
	return strings.Join(strings.Fields(root), "")
}

// NamesInDepth returns all 99, ordered by number.
func (e *Engine) NamesInDepth() []NameDepth { return e.namesDepth }

// NameInDepth returns one Name's depth entry, by its number (1..99).
func (e *Engine) NameInDepth(number int) (NameDepth, bool) {
	for _, name := range e.namesDepth {
		if name.Number == number {
			return name, true
		}
	}
	return NameDepth{}, false
}

// NameThemes returns the nine themes, in the corpus's own order.
func (e *Engine) NameThemes() []NameTheme { return e.nameThemes }

// NamesByTheme returns every Name under one theme, by number.
func (e *Engine) NamesByTheme(theme string) []NameDepth {
	out := []NameDepth{}
	for _, name := range e.namesDepth {
		if name.Theme == theme {
			out = append(out, name)
		}
	}
	return out
}

// NamesByRoot returns every Name built on one root. Accepts either spelling, spaced or closed up.
func (e *Engine) NamesByRoot(root string) []NameDepth {
	key := RootKey(root)
	if key == "" {
		return nil
	}
	out := []NameDepth{}
	for _, name := range e.namesDepth {
		if RootKey(name.Root) == key {
			out = append(out, name)
		}
	}
	return out
}

// NameInAyahHit is a Name and the occurrence that put it in an ayah.
type NameInAyahHit struct {
	Name       NameDepth
	Occurrence NameOccurrence
}

// NamesInAyah returns every Name appearing in an ayah, in token order. An unplaced occurrence
// (a nil Token) sorts last, so a highlighted list stays in reading order.
func (e *Engine) NamesInAyah(surahID, ayahID int) []NameInAyahHit {
	hits := []NameInAyahHit{}
	for _, name := range e.namesDepth {
		for _, occurrence := range name.Occurrences {
			if occurrence.Surah == surahID && occurrence.Ayah == ayahID {
				hits = append(hits, NameInAyahHit{Name: name, Occurrence: occurrence})
			}
		}
	}
	at := func(o NameOccurrence) int {
		if o.Token == nil {
			return int(^uint(0) >> 1)
		}
		return *o.Token
	}
	sort.SliceStable(hits, func(i, j int) bool { return at(hits[i].Occurrence) < at(hits[j].Occurrence) })
	return hits
}

// SearchNamesInDepth matches the root (spaces closed on both sides), the explanation and the
// living line.
func (e *Engine) SearchNamesInDepth(query string, limit int) []NameDepth {
	q := strings.TrimSpace(query)
	if q == "" {
		return nil
	}
	lower := strings.ToLower(q)
	key := RootKey(q)
	out := []NameDepth{}
	for _, name := range e.namesDepth {
		if (key != "" && strings.Contains(RootKey(name.Root), key)) ||
			strings.Contains(strings.ToLower(name.Explanation), lower) ||
			strings.Contains(strings.ToLower(name.Living), lower) {
			out = append(out, name)
			if limit > 0 && len(out) >= limit {
				break
			}
		}
	}
	return out
}

// NamesDepthCount reports how many Names carry depth, how many themes, and how many occurrences
// they cover.
func (e *Engine) NamesDepthCount() (names, themes, occurrences int) {
	for _, name := range e.namesDepth {
		occurrences += len(name.Occurrences)
	}
	return len(e.namesDepth), len(e.nameThemes), occurrences
}

// ---- the chains of transmission --------------------------------------------------

// IsnadNode is one person in a chain.
type IsnadNode struct {
	Name   string `json:"name"`
	Arabic string `json:"arabic"`
	// Detail is the death year, e.g. "d. 117 AH".
	Detail string `json:"detail"`
	Role   string `json:"role"`
}

// IsnadLayer is one generation of a chain, as a consumer draws it: a row, connected downward.
type IsnadLayer struct {
	Title string
	Nodes []IsnadNode
}

// ImamChain is an imam's side: the Successors he read on, and the Companions they read on.
type ImamChain struct {
	Teachers   []IsnadNode `json:"teachers"`
	Companions []IsnadNode `json:"companions"`
}

// NarratorChain is a narrator's side: the links between him and the imam (empty where he read on
// the imam himself), and the students who carried his narration on.
type NarratorChain struct {
	// Imam is the imam this narration comes from.
	Imam     string      `json:"imam"`
	Links    []IsnadNode `json:"links"`
	Students []IsnadNode `json:"students"`
}

type isnadFile struct {
	Prophet    *IsnadNode               `json:"prophet"`
	Companions []IsnadNode              `json:"companions"`
	Imams      map[string]ImamChain     `json:"imams"`
	Narrators  map[string]NarratorChain `json:"narrators"`
}

// IsnadProphet returns the head of every chain.
func (e *Engine) IsnadProphet() (IsnadNode, bool) {
	if e.isnad.Prophet == nil {
		return IsnadNode{}, false
	}
	return *e.isnad.Prophet, true
}

// IsnadCompanions returns the thirteen Companions the readings are transmitted from.
func (e *Engine) IsnadCompanions() []IsnadNode { return e.isnad.Companions }

// IsnadImamKeys returns the ten imams' keys, sorted.
func (e *Engine) IsnadImamKeys() []string {
	keys := make([]string, 0, len(e.isnad.Imams))
	for key := range e.isnad.Imams {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}

// IsnadNarratorKeys returns the twenty riwayah tags, sorted.
func (e *Engine) IsnadNarratorKeys() []string {
	keys := make([]string, 0, len(e.isnad.Narrators))
	for key := range e.isnad.Narrators {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	return keys
}

// IsnadImam returns one imam's side of the chain.
func (e *Engine) IsnadImam(imam string) (ImamChain, bool) {
	chain, ok := e.isnad.Imams[imam]
	return chain, ok
}

// IsnadNarrator returns one narrator's side of the chain.
func (e *Engine) IsnadNarrator(riwayah string) (NarratorChain, bool) {
	chain, ok := e.isnad.Narrators[riwayah]
	return chain, ok
}

// IsnadReadsDirectly reports whether a narrator read on his imam himself, with nobody between.
func (e *Engine) IsnadReadsDirectly(riwayah string) bool {
	chain, ok := e.isnad.Narrators[riwayah]
	return ok && len(chain.Links) == 0
}

// IsnadImamOf returns the imam a riwayah comes from.
//
// Read from the corpus, NOT parsed off the tag: four tags name the imam in the Arabic genitive
// ("ad-Duri an Abi Amr") while his key is the nominative ("Abu Amr"), so splitting on " an "
// would resolve those four to nothing.
func (e *Engine) IsnadImamOf(riwayah string) (string, bool) {
	chain, ok := e.isnad.Narrators[riwayah]
	if !ok {
		return "", false
	}
	if _, known := e.isnad.Imams[chain.Imam]; !known {
		return "", false
	}
	return chain.Imam, true
}

func isnadNarratorName(riwayah string) string {
	if i := strings.Index(riwayah, " an "); i >= 0 {
		return riwayah[:i]
	}
	return riwayah
}

func isnadPlain(name, role string) IsnadNode {
	return IsnadNode{Name: name, Role: role}
}

// IsnadTopLayers returns the layers above any imam: the Prophet, the Companions his teachers read
// on, and those teachers. Shared by both chain forms, since every chain runs through them.
func (e *Engine) IsnadTopLayers(imam string) []IsnadLayer {
	chain, ok := e.isnad.Imams[imam]
	if !ok {
		return nil
	}
	layers := []IsnadLayer{}
	if e.isnad.Prophet != nil {
		layers = append(layers, IsnadLayer{Title: "THE PROPHET", Nodes: []IsnadNode{*e.isnad.Prophet}})
	}
	if len(chain.Companions) > 0 {
		layers = append(layers, IsnadLayer{Title: "THE COMPANIONS", Nodes: chain.Companions})
	}
	if len(chain.Teachers) > 0 {
		title := "HIS TEACHERS"
		if len(chain.Teachers) == 1 {
			title = "HIS TEACHER"
		}
		layers = append(layers, IsnadLayer{Title: title, Nodes: chain.Teachers})
	}
	return layers
}

// IsnadChain returns a whole chain as layers: a riwayah tag for one narration's chain, or an imam
// key for the reading's, which ends at his two narrators.
func (e *Engine) IsnadChain(key string) []IsnadLayer {
	key = strings.TrimSpace(key)
	if _, ok := e.isnad.Imams[key]; ok {
		layers := e.IsnadTopLayers(key)
		if len(layers) == 0 {
			return layers
		}
		layers = append(layers, IsnadLayer{Title: "THE IMAM", Nodes: []IsnadNode{isnadPlain(key, "imam")}})
		nodes := []IsnadNode{}
		for _, tag := range e.IsnadNarratorKeys() {
			if imam, ok := e.IsnadImamOf(tag); ok && imam == key {
				nodes = append(nodes, isnadPlain(isnadNarratorName(tag), "narrator"))
			}
		}
		if len(nodes) > 0 {
			layers = append(layers, IsnadLayer{Title: "HIS TWO NARRATORS", Nodes: nodes})
		}
		return layers
	}

	narrator, ok := e.isnad.Narrators[key]
	if !ok {
		return nil
	}
	imam, ok := e.IsnadImamOf(key)
	if !ok {
		return nil
	}
	layers := e.IsnadTopLayers(imam)
	if len(layers) == 0 {
		return layers
	}
	layers = append(layers, IsnadLayer{Title: "THE IMAM", Nodes: []IsnadNode{isnadPlain(imam, "imam")}})
	if len(narrator.Links) > 0 {
		title := "THE LINKS BETWEEN"
		if len(narrator.Links) == 1 {
			title = "THE LINK BETWEEN"
		}
		layers = append(layers, IsnadLayer{Title: title, Nodes: narrator.Links})
	}
	layers = append(layers, IsnadLayer{Title: "THE NARRATOR", Nodes: []IsnadNode{isnadPlain(isnadNarratorName(key), "narrator")}})
	if len(narrator.Students) > 0 {
		layers = append(layers, IsnadLayer{Title: "HIS STUDENTS", Nodes: narrator.Students})
	}
	return layers
}

// IsnadSentence returns one sentence on how a narrator reaches his imam: directly, or through the
// links between.
func (e *Engine) IsnadSentence(riwayah string) string {
	chain, ok := e.isnad.Narrators[riwayah]
	if !ok {
		return ""
	}
	imam, ok := e.IsnadImamOf(riwayah)
	if !ok {
		return ""
	}
	narrator := isnadNarratorName(riwayah)
	if len(chain.Links) == 0 {
		return narrator + " read on " + imam + " himself, and " + imam +
			"'s chain runs through his teachers to the Companions and to the Prophet ﷺ."
	}
	names := make([]string, 0, len(chain.Links))
	for _, node := range chain.Links {
		names = append(names, node.Name)
	}
	path := names[0]
	if len(names) > 1 {
		path = strings.Join(names[:len(names)-1], ", ") + " and then " + names[len(names)-1]
	}
	return narrator + " did not meet " + imam + ": the reading reached him through " + path +
		", and from " + imam + " it runs through his teachers to the Companions and to the Prophet ﷺ."
}

// IsnadCount reports how many imams, narrators and Companions the chains cover.
func (e *Engine) IsnadCount() (imams, narrators, companions int) {
	return len(e.isnad.Imams), len(e.isnad.Narrators), len(e.isnad.Companions)
}

File: packages/quran-engine-rust/tests/batch6.rs
//! Batch 6: the 99 Names in depth, and the chains of transmission of the Ten Readings.
//!
//! A deliberate translation of packages/quran-engine-js/test/batch6.test.js: the same corpora,
//! the same counts, the same shape checks, so a port that drifts fails here rather than in a
//! consumer.

use std::path::Path;

use quran_engine::{batch6::root_key, Engine};

fn engine() -> Engine {
    Engine::load(Path::new("../../data")).expect("load engine")
}

fn fold(text: &str) -> String {
    text.chars()
        .filter(|c| {
            !matches!(*c as u32, 0x064B..=0x065F | 0x0670 | 0x06D6..=0x06ED | 0x0640)
        })
        .map(|c| match c {
            '\u{0671}' | '\u{0623}' | '\u{0625}' => '\u{0627}',
            '\u{0624}' => '\u{0648}',
            '\u{0626}' => '\u{064A}',
            other => other,
        })
        .collect()
}

// ---- the Names in depth ----------------------------------------------------------

#[test]
fn names_depth_counts() {
    let e = engine();
    assert_eq!(e.names_depth_count(), (99, 9, 194));
    let numbers: Vec<u32> = e.names_in_depth().iter().map(|n| n.number).collect();
    assert_eq!(numbers, (1..=99).collect::<Vec<u32>>());
}

#[test]
fn every_name_has_root_theme_and_prose() {
    let e = engine();
    let ids: Vec<&str> = e.name_themes().iter().map(|t| t.id.as_str()).collect();
    for name in e.names_in_depth() {
        assert!(!name.root.is_empty(), "name {} has no root", name.number);
        assert!(ids.contains(&name.theme.as_str()), "name {} theme {}", name.number, name.theme);
        assert!(name.explanation.len() > 40, "name {} explanation", name.number);
        assert!(name.living.len() > 20, "name {} living", name.number);
    }
}

#[test]
fn themes_partition_the_ninety_nine() {
    let e = engine();
    let mut total = 0;
    for theme in e.name_themes() {
        let members = e.names_by_theme(&theme.id);
        assert!(!members.is_empty(), "theme {} has no Names", theme.id);
        total += members.len();
    }
    assert_eq!(total, 99);
}

#[test]
fn a_root_is_spaced_and_closes_up() {
    let e = engine();
    let rahman = e.name_in_depth(1).expect("name 1");
    assert_eq!(rahman.root, "ر ح م");
    assert_eq!(root_key(&rahman.root), "رحم");
    let by_closed: Vec<u32> = e.names_by_root("رحم").iter().map(|n| n.number).collect();
    let by_spaced: Vec<u32> = e.names_by_root("ر ح م").iter().map(|n| n.number).collect();
    assert_eq!(by_closed, vec![1, 2]);
    assert_eq!(by_spaced, vec![1, 2]);
}

#[test]
fn placed_occurrences_are_real_tokens() {
    let e = engine();
    let (mut placed, mut unplaced) = (0, 0);
    for name in e.names_in_depth() {
        for o in &name.occurrences {
            let ayah = e
                .ayah(o.surah, o.ayah)
                .unwrap_or_else(|| panic!("name {} points at {}:{}", name.number, o.surah, o.ayah));
            match o.token {
                None => {
                    assert_eq!(o.tokens, 0, "name {} unplaced but spans {}", name.number, o.tokens);
                    unplaced += 1;
                }
                Some(token) => {
                    let tokens: Vec<&str> = ayah.text_arabic.split_whitespace().collect();
                    assert!(token < tokens.len(), "name {} token {} past the end", name.number, token);
                    assert!(!fold(tokens[token]).is_empty());
                    placed += 1;
                }
            }
        }
    }
    assert_eq!(placed, 184);
    assert_eq!(unplaced, 10);
}

#[test]
fn the_basmalah_carries_two_names() {
    let e = engine();
    let hits = e.names_in_ayah(1, 3);
    let numbers: Vec<u32> = hits.iter().map(|(n, _)| n.number).collect();
    let tokens: Vec<Option<usize>> = hits.iter().map(|(_, o)| o.token).collect();
    assert_eq!(numbers, vec![1, 2]);
    assert_eq!(tokens, vec![Some(0), Some(1)]);
}

#[test]
fn names_search_by_root_and_prose() {
    let e = engine();
    assert!(e.search_names_in_depth("رحم", 25).iter().any(|n| n.number == 1));
    let first = e.name_in_depth(1).expect("name 1");
    let word = first.living.split_whitespace().find(|w| w.len() > 6).expect("a long word");
    assert!(!e.search_names_in_depth(word, 25).is_empty());
    assert!(e.search_names_in_depth("", 25).is_empty());
}

// ---- the chains of transmission --------------------------------------------------

#[test]
fn isnad_counts() {
    let e = engine();
    assert_eq!(e.isnad_count(), (10, 20, 13));
    assert!(e.isnad_prophet().is_some());
}

#[test]
fn every_riwayah_resolves_to_an_imam_two_apiece() {
    let e = engine();
    for imam in e.isnad_imam_keys() {
        let n = e
            .isnad_narrator_keys()
            .iter()
            .filter(|tag| e.isnad_imam_of(tag).as_deref() == Some(imam))
            .count();
        assert_eq!(n, 2, "{imam} has {n} narrators");
    }
    for tag in e.isnad_narrator_keys() {
        assert!(e.isnad_imam_of(tag).is_some(), "{tag} resolves to no imam");
    }
}

#[test]
fn every_imam_reaches_companions() {
    let e = engine();
    for imam in e.isnad_imam_keys() {
        let chain = e.isnad_imam(imam).expect("imam chain");
        // Abu Jafar WAS a Successor and read on Companions himself, so he has no teachers layer.
        if imam != "Abu Jafar" {
            assert!(!chain.teachers.is_empty(), "{imam} has no teachers");
        }
        assert!(!chain.companions.is_empty(), "{imam} reaches no Companion");
        for node in &chain.teachers {
            assert_eq!(node.role, "successor");
            assert!(node.detail.starts_with("d."), "{} has no death year", node.name);
        }
        for node in &chain.companions {
            assert_eq!(node.role, "companion");
        }
    }
}

#[test]
fn a_chain_runs_from_the_prophet_to_the_narrator() {
    let e = engine();
    for tag in e.isnad_narrator_keys() {
        let layers = e.isnad_chain(tag);
        assert!(layers.len() >= 4, "{tag} chain is only {} layers", layers.len());
        assert_eq!(layers[0].title, "THE PROPHET");
        assert_eq!(layers[1].title, "THE COMPANIONS");
        let titles: Vec<&str> = layers.iter().map(|l| l.title.as_str()).collect();
        let imam = titles.iter().position(|t| *t == "THE IMAM").expect("an imam layer");
        let narrator = titles.iter().position(|t| *t == "THE NARRATOR").expect("a narrator layer");
        assert!(imam < narrator, "{tag} has the narrator above the imam");
    }
}

#[test]
fn an_imam_chain_ends_at_his_two_narrators() {
    let e = engine();
    for imam in e.isnad_imam_keys() {
        let layers = e.isnad_chain(imam);
        let last = layers.last().expect("a last layer");
        assert_eq!(last.title, "HIS TWO NARRATORS", "{imam}");
        assert_eq!(last.nodes.len(), 2, "{imam}");
    }
}

#[test]
fn reading_directly_is_having_no_links() {
    let e = engine();
    for tag in e.isnad_narrator_keys() {
        let chain = e.isnad_narrator(tag).expect("narrator chain");
        assert_eq!(e.isnad_reads_directly(tag), chain.links.is_empty(), "{tag}");
        let sentence = e.isnad_sentence(tag);
        assert!(!sentence.is_empty(), "{tag}");
        assert_eq!(sentence.contains("did not meet"), !chain.links.is_empty(), "{tag}");
    }
}

#[test]
fn hafs_reads_directly_and_qunbul_does_not() {
    let e = engine();
    assert_eq!(e.isnad_imam_of("Hafs an Asim").as_deref(), Some("Asim"));
    assert!(e.isnad_reads_directly("Hafs an Asim"));
    assert!(e.isnad_sentence("Hafs an Asim").starts_with("Hafs read on Asim himself"));

    assert_eq!(e.isnad_imam_of("Qunbul an Ibn Kathir").as_deref(), Some("Ibn Kathir"));
    assert!(!e.isnad_reads_directly("Qunbul an Ibn Kathir"));
    assert_eq!(e.isnad_narrator("Qunbul an Ibn Kathir").expect("chain").links.len(), 3);
}

#[test]
fn an_unknown_key_answers_nothing() {
    let e = engine();
    assert!(e.isnad_chain("Nobody an Nobody").is_empty());
    assert_eq!(e.isnad_sentence("Nobody an Nobody"), "");
    assert!(e.isnad_imam_of("no separator here").is_none());
}
